server/command/command.go

```diff
@@ -0,0 +1,110 @@
+package command
+
+import (
+	"encoding/json"
+
+	"github.com/mattermost/mattermost-plugin-todo/server/todo"
+	"github.com/mattermost/mattermost-server/v5/model"
+)
+
+type Command struct {
+	Args      *model.CommandArgs
+	Log       logger
+	pluginAPI pluginAPI
+	todoAPI   todoAPI
+}
+
+type ResponseType string
+
+// ResponseType values
+const (
+	ResponseTypeEphemeral    ResponseType = "ephemeral"
+	ResponseTypePost         ResponseType = "post"
+	ResponseTypeBotCustomDM  ResponseType = "bot_custom_dm"
+	ResponseTypeBotReplyPost ResponseType = "bot_reply_post"
+	ResponseTypeBotDM        ResponseType = "bot_dm"
+	ResponseTypePayload      ResponseType = "payload"
+)
+
+type Response struct {
+	Type      ResponseType
+	Message   string
+	Todo      string
+	UserID    string
+	ChannelID string
+	IssueID   string
+	PostID    string
+}
+
+type logger interface {
+	LogError(msg string, keyValuePairs ...interface{})
+}
+
+type pluginAPI interface {
+	GetUserByUsername(name string) (*model.User, *model.AppError)
+	SendRefreshEvent(userID string)
+}
+
+type todoAPI interface {
```

Comment:
These apis are defined also on service almost the same. Any strong reason to have declared twice?

Reply:
Thats an excellent question. I went for this style, idea is to declare interfaces where they are needed. Package declares an interface which it wants to receive from other package(not other way around). Function's arguments are interfaces and  return types are concrete. This way unit testing is easier. Downside is we might end up with more interfaces in the code. I wanted everyones input on this.

Reply:
Here's where I arrived to after my own evolution. Basically all "APIs" should be expressed as a struct in, struct+error out. `out` implements a `json.Marshaler` and `md.Markdowner` (+ other "marshaling" interfaces if necessary, for websocket updates and such).

As to the dogma of where interfaces should be declared... I see nothing wrong with a package declaring its own interfaces. If the client wants to use this "standard" definition - it can. Otherwise it can define its own interfaces - all of the signature matching works anyway. 0/5 on this one, but I feel that specifically, `server/todo` should declare its interface for mocking.

Reply:
0/5 but I like interfaces to be declared just once. Whenever there is a change on an interface, it will gracefully propagate to all the places it is needed, having to do a bit less of changes here and there. For example, if I declare a Poster that implements `DM(user, message) error` and then I need to return the post ID, and change it to `DM(user, message) (postID, error)`, I would need to make that changes on all the different interfaces, potentially hiding some errors until later.

Maybe I am thinking too much and there are not real cases like this, but it is something that worry me.

Reply:
I'd suggest to introduce exported interfaces when we need them and not before. In general I agree that in future `todo` package will need an interface and mocks as well.